### ISSUE TITLE
Consider using leading zeros in OSCAL Content Identifiers

### ISSUE THREAD

**Issue by david-waltermire-nist** [usnistgov/OSCAL#349]
# User Story:

As an OSCAL content consumer, I need a reliable method to sort controls represented in an OSCAL catalog format. Adding zeros to the control (and possibly other) identifiers will provide content users with a simple method to sort a listing of controls using dictionary ordering.

## Goals:

Add leading zeros to the identifiers in the OSCAL catalog examples provided by NIST and FedRAMP in the OSCAL GitHub repository.

## Dependencies:

None.

## Acceptance Criteria

1. All areas where leading zeros need to be applied to identifiers have been documented in this issue.
1. All identifiers that are documented have been enhanced to include leading zeros.

**Comment by anweiss** [comment]
@david-waltermire-nist would this be somewhat dependent on #280?

**Comment by trevor-vaughan** [comment]
I don't think I would do this since the number of leading zeroes would be completely dependent on the number of items in the content (if I'm reading things right).

Human sort is easy enough to do in pretty much every programming language.

**Comment by iMichaela** [comment]
## 5/23/2019
will be done at the same time with #346 . @wendellpiez needs to discuss with @david-waltermire-nist to finalize the design.

**Comment by wendellpiez** [comment]
An XSLT that implements a design is in my kit (`id-expand.xsl`). It produces a copy of an OSCAL document (catalog or profile) except padding zeroes in ID values as described, according to this map:

* `ac-1`         -> `ac-01`
* `ac-2.1`       -> `ac-02.01`
* `ac-2.11`      -> `ac-02.11`
* `ac-2.1_obj`   -> `ac-02.01_obj`
* `ac-2.1_obj.2` -> `ac-02.01_obj.02`

This should provide for all the sorting capabilities wanted.

**Comment by david-waltermire-nist** [comment]
@trevor-vaughan I agree. I have been on the fence about this issue, with a leaning towards not making this change. We do need to make a choice now, as we would like to stabilize the catalog and profiles based on SP 800-53. There are many application specific ways to deal with this, so it's not the end of the world if we keep the IDs as-is.

@wendellpiez What are your thoughts on this?

**Comment by david-waltermire-nist** [comment]
Discussed this with @wendellpiez and we agreed that this would cause significant downstream impacts, with no long-term guarantee of consistency. For example, it will be hard to enforce other catalog maintainers into a specific id pattern. As a result, we agreed not to make this change.

**Comment by wendellpiez** [comment]
My feeling is that this change, if we made it, could only affect the single catalog: this would not be a policy for all catalogs. (Hence evading the question of how many zeroes since that is known in advance.) However, I am also willing to listen to those who say "it's not really a problem". To the eye, I would say it's a tradeoff - the padded IDs are slightly less legible but they are sortable.

Indeed, among the signals that catalogs send about their differentiation from one another, are their various naming and labeling schemes (commonly reflecting their different hierarchies). IMO the rule for catalogs should be "any rule works (make a rule)".

**Comment by david-waltermire-nist** [comment]
We will investigate adding some sort-ordering properties on controls to address this in a different way.

**Comment by aaronlippold** [comment]
Hello, my two cents. Fundamentally the issue here isn't the format of the
strings but the fact that we want to be able to sort things in expected
ascending order. So given the nature of integers and how a computer sorts
numbers standardizing on a single format such that sorting in ascending or
descending order produces known results is really what we need to get to
regardless of the profile.

If you don't do this your end users like me will just amend the data to
achieve this as we already have :-)

Thanks

**Comment by iMichaela** [comment]
These sort-ordering properties on controls are more dificults to design if we are targeting to support ordering all contorls coming from different catalogs (e.g. 800-53 + HIPPA + others) in one document. @aaronlippold - from your and others' experience - is this complexity justified by the frequency encounter of such scenario?

**Comment by aaronlippold** [comment]
Well, in all fairness that would be something you the project would have to
decide on balance with the rest of your requirements.

But I think this is getting more towards a normalization of the data
structure question.

 I can only tell you that as an end user when I print my document I will
want to print things in alphabetical then subsorted by numerical order as
we have in the templates for years.

Ac-01
Ac-01a
...

AC-1 is the pretty print version of AA-NNN[.][a][.][nn]

AC-1 a
AC-2.b.1
...

It terms of data storage, if you are going to use the name of the items as
the primary index that index needs to be regular and the structure above is
how a computer would sort such a structure.

I would use the above method as a secondary approach.

As a primary approach, I would use integer sequences as the true primary
index for lists of things and something's name is a secondary field. For a
data set that has an inherent order you just have to insert it in that
order and if you want to move or insert a new item into the order you know
where to put it. And when you ask it to sort things it's rather simple.

As to your second question if two sets have an inherent order you can solve
this again with the integer index.

This does add extra complexity in some of the querying statements because
you are asking the database to return items with particular indexes and
having to map them back to the human readable names of those indexes, the
raw data would would be just a string of comma separated numbers etc but
this can be streamlined through things like views and procedures.

Given that you will create a filter for display this hash of index to human
display name is an o(1) lookup which can be stored in memory.

How we display the data versus how we store it in the database do not
necessarily need to be the same. We can easily generate a filter which can
display the information in a more readable format. This is the same
consideration we have for dates in a computer. How we store dates in a
computer is completely different as how we display them to humans.

Yours

Aaron

**Comment by aaronlippold** [comment]
Another question to consider is are these sets static? I think they are. We
actually know the number of items in each of the sets. Rev4 has a no and
number of family sub families etc etc as will rev5.

But others in the list might have better suggestions on back end
implementation.

A

**Comment by wendellpiez** [comment]
For OSCAL in general, I do not believe we need to write a spec for this. Requirements are too local. And a catalog that needs to be able to dictate a sort order, can always do this.

For SP 800-53 XML, one option would be to release either an XSLT (I have one we could adapt), or even an OSCAL profile, that would tag the catalog controls with properties for more normalized identifiers. And since anyone can do this -- it doesn't seem like a big problem to me.

**Comment by brianrufgsa** [comment]
@wendellpiez - In our meeting last week, I made the same argument @aaronlippold does above about sorting using ID. In that meeting a suggestion was made to allow a prop statement in the control that has a sequence assigned to it.

I was initially resistant to this until the point was made that OSCAL also enables an organization to pull together controls from multiple compliance regimes, such as with NIH that might be subject to HIPAA as well as 800-53. The approach of having a prop statement with the order allows organizations to organize a "mixed catalog"of requirements like this as they see fit.

I can live with the prop suggestion. I am not nearly as comfortable with the XSLT suggestion. EVERYONE who adopts 800-53 will need to have it presented in the same order in which it appears in the PDF file. The XSLT suggestion forces every single adopter of 800-53 to setup an XSLT file to achieve that sequencing.

Let's stick with the prop element for sequencing as we all agreed to in that meeting.

**Comment by wendellpiez** [comment]
Brian, this is fine (no problem at all) -- except I have to push back on the idea that offering an XSLT-based solution forces everyone to use an XSLT-based solution. :wink:

**Comment by brianrufgsa** [comment]
@wendellpiez, I was more focused on the idea that the intended sequence of controls within a given catalog should be built into the catalog data itself by the publisher of that catalog. In this case 800-53 has a clear intended sequencing, that every 800-53 adopter will need the same sequencing of 800-53. 

Whether with XSLT or some other way, having each adopter establish their own sorting mechanism doesn't make sense when they will all need the same sequence from the 800-53 catalog.
I can agree with you for other desired sorting, but this particular sequence is representative of how the document was published, has an intended progression by the author, and is universal to all adopters of this catalog.

**Comment by brianrufgsa** [comment]
I want to land the plane on this issue.

First, this is an issue of whether the OSCAL syntax should support the ability for an OSCAL publisher to indicate their intended sequence of controls? 
If so, it becomes a discussion as to what mechanism(s) should be available within the OSCAL syntax to support this.

I believe there is a clear requirement for an OSCAL content publisher to specify the intended sequence. I believe we had consensus on this requirement in last week's status meeting, and we can point to several use cases where this sequence is expected/required.

Assuming there are no compelling descending opinions, I think the conversation becomes one of "how" the OSCAL syntax can support this. I see three possibilities:

**1. Relative Placement:** The sequence of the controls in the catalog are the intended order. 

Pro: It's simple and easy to implement for XML. It also supports the sorting of merged controls from multiple sources. (ie. a single catalog that includes an organization's 800-53, HIPAA, and PCI requirements.) 

Cons: I have heard JSON object models often do not respect this order, so it is unreliable in JSON format. Similarly, if the data is imported to a database, some tagging must be added to maintain this sequencing or it will be lost.

**2. Sortable IDs:**  Padding zeros to ensure a text sort of the controls by ID field results in the publisher's intended order.

Pros: Easy and straight-forward. No unnecessary duplication.

Cons: May force IDs to change in later revisions. For example, if Rev 5 requires three digit numbers instead of two, the IDs would change. While this would be a separate catalog and could work in theory, it adds a layer of complication to version transition for adopters. 

Does not support the sorting of merged controls from multiple sources. (ie. a single catalog that includes an organization's 800-53, HIPAA, and PCI requirements.) 

**3. XSL Transformation:** Have either the OSCAL content publisher or OSCAL users develop XSL Transformations to achieve the intended order. 

Pros: Allows for flexibility.

Cons: Requires the OSCAL publisher to also publish an XSLT file along with the catalog, which then requires all consumers of that catalog to use that XSLT if they require the publisher's intended sequencing.

**4. Sequence Property or Attribute:** This adds an additional property or attribute to the OSCAL syntax, which allows an OSCAL content publisher to explicitly specify an intended sequence. 

Pros: Clear and unambiguous. Supports the sorting of merged controls from multiple sources. (ie. a single catalog that includes an organization's 800-53, HIPAA, and PCI requirements.)  

Con:s Must be added to the syntax. While not necessary, the 800-53 catalog should be updated to reflect the intended sequence.

**My Recommendation:**
#4 is what the team agreed to in last Thursday's status meeting, and I still advocate strongly this is the best solution. 

It can be an optional property, and if an OSCAL content publisher elects to also use sortable IDs (Option #1), they can do so. Either way, OSCAL consumers can still use the sequence property when merging controls from multiple catalogs. 

I don't have a strong feeling as to whether we use an additional property ( <prop name="sequence" /> ) or an attribute ( <control class="SP800-53" id="ac-1" sequence="ac-01"> )

We would need to state our position regarding controls vs. subcontrols. While I anticipate catalog publishers would always want subcontrols to remain under their associated control, I can see organizations wanting to organize subcontrols differently, thus I think any sequence tag should be honored (when sorted by sequence), even if it means a subcontrols appears in a location other than with its associated control. 

**One Final Thought**
The NIST OSCAL team is responsible for defining the OSCAL syntax. This team has developed a copy of NIST SP 800-53 as an OSCAL catalog, but ultimately the NIST FISMA team is responsible for 800-53 content. I don't know what agreements the NIST OSCAL Team has made with the NIST FISMA Team in this regard, but it seems to me we are debating content points (ie padding the IDs) that they should at least have an opportunity to weigh in on.

**Comment by wendellpiez** [comment]
@brianrufgsa I was hoping that if it came down to writing a literal value, we could do something like

```xml
<prop name="sort-key">AC-02.02</prop>
```

In passing, I note:

1. As someone has remarked I think, this is a redundant value from the point of view of the information model, if taken as a whole (that is, including its specs). Given prior knowledge of how the controls are produced, they are easily sortable even in the worst case. We could even write up the pattern in documentation and leave it at that. (However, as has been observed, this pattern will be unique to the catalog or family of catalogs.)

2. Of course worst case is not the most usual case, and much of the whole point of the XML model is the respect for ordering of arbitrary elements in this sort of situation, so your answer (1) above is the correct answer from the XML point of view. Indeed, this need to respect ordering in the original data is also why we have an "as-is" option in the OSCAL profile model - it is intended to provide for both the ordering and the grouping of controls in the profile as they are ordered and grouped in the catalog(s) from which they are imported. (We haven't decided whether a conformant application will be expected to be able to do this or if it would be best left optional. But we have thought about how to express it.)

    Now as you say there is a complication because the "it's always ordered" rule does not hold in JSON - except that, in OSCAL JSON we have actually taken care to cast to arrays in the general case of modeling (what are in XML) element contents, precisely so that we can order things. C'est accompli. Except for the complication of what happens if you promote IDs to be labels in your JSON. Then what you had before as an array, is a set of labeled objects, and we are back to square one and indeed AC-10 will frequently come before AC-2 without further intervention (a problem), if you have used the 'ac-10' and 'ac-2' ids for your keys. (Maybe a reason not to use this Metaschema feature after all?)

3. What further intervention could be possible in this drastic case where ordering fell out of sync or at any rate could not be trusted? A couple of possibilities come to mind:

a. Refer back to the original spec (perhaps a trusted or canonical catalog?) and try to order by reference to that
b. Derive a "correct order" based on prior or acquired knowledge about the ordering (in this case, that padding numerals in IDs to two places is sufficient)
c. Run someone else's code to do this (XSLT or other) and check the results

There might be other possibilities I haven't thought of yet. It is true, Probably none of them is as easy as looking at the data.

I am not offering a vote here, merely information I think is relevant (and might answer questions you pose).

**Comment by david-waltermire-nist** [comment]
@brianrufgsa I like your summary of the options. As we discussed when we met last week, I believe option 4 is our best route considering all the options.

One thing to note, it is that we are adding the ids being discussed into the OSCAL versions of the FISMA content. These IDs do not appear in the source documents. They are an invention in the OSCAL content to allow more fine-grained references to identified content.

As for the FISMA team. We are working with them to produce OSCALized versions of the SP 800-53 content. They know the subject matter well, but we know OSCAL well. Any changes to the FISMA OSCAL content will need to come from us, with review from them.

**Comment by wendellpiez** [comment]
Since this didn't make it into M1, maybe we should consider another of the options, namely publishing the nominal sort order as an OSCAL profile?

I suggest this not in order to counter #4 (perfectly good solution) but only to remind ourselves that many things are possible - as a profile, we could do this right away for rev4 as stipulated. The profile can be machine-produced.

**Comment by brianrufgsa** [comment]
It sounds like we've reached consensus about the existence of a sort-order field, so a catalog publisher can indicate the intended sequencing of controls; however, we have not yet reached consensus on how to populate that field.

On that second point it is important to clarify that "how" to populate that field is going to be specific to each compliance regime, so for NIST's purposes, we are only discussing how to populate that field for 800-53, Revision 4. The topic _will_ need to be revisited for by each catalog publisher, and _may_ need to be revisited by NIST when Rev 5 is published.

For 800-53, Rev 4 - I've heard a few options, and see a few more:
- **Zero Padding**: As discussed, use all lower case alphabetic characters, and zero-pad all numbers to consistently show two digits. (eg "ac-02", "ac-02.01", "ac-02.02")
- Use **straight integers** and simply number the controls/sub-controls. (eg "1", "2", "3")
- Use **integers for controls, and decimals for sub-controls**, which would still require the sub-controls to have padded zeros for numeric sorting (eg 2, 2.01, 2.02)

I can think of several other variations, but don't see the value in an exhaustive list. Does anyone have another variation we should strongly consider before we make a decision?

**Comment by georgalis** [comment]
I'll comment that timestamp based versioning has been very effective for me. Since sequential numbers cannot even be use to count releases, they only point to a qualifying index, using date based release numbers adds meaningful information to the name. In some cases I base on ISO 8601, eg ac-20190624 (the most significant 8 characters), but generally I use unix seconds, represented in hexadecimal, eg ac-5d110206, which also uses 8 characters but helpful for indexing files that may have multiple releases per day, for the foreseeable future. It sorts and in practice you can use a variation of scientific notation that disregards the exponent and uses only as many significant digits required for a unique identifier: ac-5d110206, ac-5d1107b3 and ac-5d11100e becomes ac-5, ac-5d1107 and ac-5d111. While I record the full unix seconds  (and tai64n timestamps) in document meta data, it is unusual to require more than 4 or 5 characters, in what I call an epoch, the most significant unique hexadecimal unix seconds, for the record, in the domain, which 
- sorts
- is very easy for human transcription or comparison
- and, for major documentation types, typically will require only 3 or 4 characters, eg ac-5d1.

The system can also be very effective as short primary key, for example this document as 5d1-ac could live in the same index, chronologically between 5d0-aX and a forthcoming document 5d11-aY.
- 5d0-aX
- 5d1-ac
- 5d11-aY

or to put another way, pad the revision with only the required significant digits of the current time, for a short, unique representation, which is sortable and indicates approximate acquisition time, when the time exponent is inferred from missing digits.

**Comment by georgalis** [comment]
I missed a nuance in my above comment, but the epoch system could certainly complement a  "Sequence Property or Attribute" for controls that are not intended for acquisition ordering. I use epochs to identify documents not revisions, so it's not unusual for me to list document epochs with their respective titles, in last modification order...

**Comment by david-waltermire-nist** [comment]
We need to add a property in the 800-53 catalog with a sortable value and update the catalog appropriately. We need to meet to design the approach and create a new issue to complete this work.

**Comment by wendellpiez** [comment]
Here is a bit of XSLT that can convert a string containing numbers into its corresponding string with those same numbers, except zero-padded (to two places). So "a1-1b" becomes "a01-01b".

```xml
<xsl:function name="o:expand" as="xs:string">
    <xsl:param name="in" as="xs:string"/>
    <xsl:value-of>
        <xsl:analyze-string select="$in" regex="\d+">
            <xsl:matching-substring>
                <xsl:value-of select="format-number(number(.),'00')"/>
            </xsl:matching-substring>
            <xsl:non-matching-substring>
                <xsl:value-of select="."/>
            </xsl:non-matching-substring>
        </xsl:analyze-string>
    </xsl:value-of>
</xsl:function>
```

We could use the above and add to every control (including enhancements) in the SP800-53 catalog a property:

```xml
<prop name="sort-id">{ o:expand($id) }</prop>
```

So we will see `<prop name="sort-id">ac-01</prop>` etc.

? Dunno if this works for a spec @brianrufgsa @david-waltermire-nist

**Comment by wendellpiez** [comment]
Amended catalog now committed to working branch https://github.com/wendellpiez/OSCAL/tree/issue-468-364-478-integration with PR to the same branch  in your repo, @david-waltermire-nist 

The new file is not much bigger than the old one at least by today's standards (less than 50KB difference).

**Comment by wendellpiez** [comment]
Proposed XML version of catalog is here: https://github.com/wendellpiez/OSCAL/blob/issue-468-364-478-integration/src/content/nist.gov/SP800-53/rev4/xml/NIST_SP-800-53_rev4_catalog.xml

Look for `<prop name="sort-id">...</prop>`.

**Comment by brianrufgsa** [comment]
@wendellpiez to provide link to latest for review. @brianrufgsa to perform peer review.

**Comment by aaronlippold** [comment]
Did we come to an agreement on this for a standardized format?

**Comment by georgalis** [comment]
I cannot speak to integration with XSLT, but I wanted to add a few additional comments about the viability of epochs, the use of "minimum most significant digits" of acquisition time stamps to create a unique identifiers per document per domain. Being able to sort an index by acquisition, does provide meaningful data, such as identifying oldest foundation documents, or recent activity.

More importantly, I'd like to question the utility of sortable identifiers to represent a hierarchical dependency tree. Is that something you actually want to do? In practice I expect a high likelihood of interpretation and/or typography errors as the documents are revised. How are errors caught and a published catalog reliably validated after insertion, edits or deletions are made to "zero padded sortable lists?"

Notable, this task is very similar to the task of operating system service start sequencing, and I've used rcorder, a tool for that, in my comments below, see https://netbsd.gw.com/cgi-bin/man-cgi?rcorder+8.i386+NetBSD-8.1

If I understand correctly, the key functionality is a means to identify "requirements" and "provides" from documents sorted in arbitrary order and provide straightforward human intrepretation. As you would expect any computer system validation documentation system to provide a format specification document, to include such fields, that would be the optimal place to identify the dependency graph and generate catalog data. For an example, I've created files with the following names

```
    5d9d 5d9d7 5d9df 5d9e 5d9e1 5d9e3 5d9e4 5d9ed 5d9f 5da 5da2 5da3 5da7
    5da78 5da7b 5da8 5da9 5daa 5dac 5dad 5dae 5dae0 5dae3 5dae4 5daf 5db
    5e4 5daa
```
[catalog-4aYs.zip](https://github.com/usnistgov/OSCAL/files/3769651/catalog-4aYs.zip)

each containing a provides field of its own name, and arbitrary references to other files, in the required field, like this:

```
==> 5da <==
# REQUIRE: 5d9df 5d9e 5dae3
# BEFORE:
# PROVIDE: 5da
# KEYWORD:

==> 5dae3 <==
# REQUIRE: 5da7
# BEFORE:
# PROVIDE: 5dae3
# KEYWORD:
```

This data is used to determine the sequence and generate a catalog entry with the meaningful reference to each file's requirements.

For catalog presentation, compose a hyphen delimited concatenation of the requirement values, appended to each file name, and of course order the list according the needful sequence. The resulting list provides the needful information for human interpretation of sequence and opportunities for parallel sequencing. In the following example it is clear seen that 5d9e1, 5da2, and 5daf could be executed first (although they are not at the beginning of sequence), and opportunity to commence any of the other documents in parallel, is readily determined, if the reader is privy to requirements already met.

```
generate_graph () {
rcorder * | while read a ; do printf '%6s' "$a"
    { sed -e '/REQUIRE/!d' -e 's/.*://' $a ;} \
        | tr ' ' '\n' | sort -u | tr '\n' '-' | sed -e 's/-$//'
    done
}
```
output:

```
rcorder: requirement `5e3' in file `5e4' has no providers.
rcorder: Circular dependency on provision `5d9e' in file `5da'.
   5e4-5e3
  5daa-5e4
  5dad-5e4
  5dac-5daa-5e4
  5d9d-5daa-5dac-5dad
 5d9df-5d9d
 5d9e1
 5da78-5d9e1-5daa-5e4
 5da7b-5da78
 5d9d7-5da7b-5daa
 5d9e3-5d9d7-5d9df
  5da3-5d9e3
  5da2
  5da7-5da2-5da3
 5dae3-5da7
   5da-5d9df-5d9e-5dae3
 5d9ed-5d9e3
  5d9f-5d9ed
  5da8-5d9ed-5d9f-5da
  5da9-5da8
  5d9e-5da8-5da9-5daa
   5db-5d9df-5d9e
  5daf
 5dae4-5d9d-5d9d7
 5d9e4-5da7b-5da8-5da9
 5dae0-5d9e4-5d9ed-5d9f-5daa-5e4
  5dae-5d9d7-5da7b-5da8-5da9
```

For each of the 27 files sorted according to dependency hierarchy (or arbitrary order when requirements don't mandate it), the first field identifies the specific file, and the remaining hyphen delimited indices reference respective requirements.

Any single document with a very large number of requirements could be problematic in this catalog format, in that case, I would suggest the "very large number of requirements" is the problem and should be addressed by consolidating into manageable sets---for the same reason we don't create master documents that require "everything"---too difficult to validate and re-validate.

The errors above very succinctly identify missing referenced documents and circular dependencies, a validation task that should be tightly coupled with any catalog generation.

Best regards,
-George

**Comment by wendellpiez** [comment]
@georgalis thanks for this, it is very interesting. I agree that some or much of the time, the simple presence of sortable labels in the underlying catalog will not be adequate even for their ostensible stated purpose. We added them mainly because the counter-argument was made that much or most of the time, they could be useful for something, while they also don't prevent more comprehensive or robust solutions, such as you have outlined.

**Comment by david-waltermire-nist** [comment]
We added the `sort-id` field in PR #492. This should provide some minimal support for sorting, although as @georgalis and @wendellpiez point out, many sorting use cases will be application specific.
